ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,1439 @@
+      }
+    }
+  };
+
+  const handleStepMove = (stepIndex: number, direction: 'up' | 'down', steps: MacroStep[]) => {
+    const newSteps = [...steps];
+    const newIndex = direction === 'up' ? stepIndex - 1 : stepIndex + 1;
+    [newSteps[stepIndex], newSteps[newIndex]] = [newSteps[newIndex], newSteps[stepIndex]];
+    return newSteps;
+  };
+
+  useEffect(() => {
+    const handleResize = () => {
+      setIsDesktop(window.innerWidth >= 768);
+    };
+    
+    window.addEventListener('resize', handleResize);
+    return () => window.removeEventListener('resize', handleResize);
+  }, []);
+
+  const { handleTouchStart, handleTouchMove, handleTouchEnd } = useTouchSort(
+    macros,
+    async (newMacros) => {
+      const updatedMacros = normalizeSortOrders(newMacros);
+      try {
+        await saveMacros(updatedMacros);
+        notifications.success("Macro order updated successfully");
+      } catch (error) {
+        if (error instanceof Error) {
+          notifications.error(`Failed to reorder macros: ${error.message}`);
+          showTemporaryError(error.message);
+        } else {
+          notifications.error("Failed to reorder macros");
+          showTemporaryError("Failed to save reordered macros");
+        }
+      }
+    }
+  );
+
+  const [showClearConfirm, setShowClearConfirm] = useState(false);
+  const [showAddMacro, setShowAddMacro] = useState(false);
+
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
+      if (e.key === 'Escape' && editingMacro) {
+        setEditingMacro(null);
+        setErrors({});
+      }
+      if (e.key === 'Enter' && (e.metaKey || e.ctrlKey)) {
+        if (editingMacro) {
+          handleUpdateMacro();
+        } else if (!isMaxMacrosReached) {
+          handleAddMacro();
+        }
+      }
+    };
+
+    window.addEventListener('keydown', handleKeyDown);
+    return () => window.removeEventListener('keydown', handleKeyDown);
+  }, [editingMacro, isMaxMacrosReached, handleAddMacro, handleUpdateMacro]);
+
+  const handleModifierChange = (stepIndex: number, modifiers: string[]) => {
+    if (editingMacro) {
+      const newSteps = [...editingMacro.steps];
+      newSteps[stepIndex].modifiers = modifiers;
+      setEditingMacro({ ...editingMacro, steps: newSteps });
+    } else {
+      const newSteps = [...(newMacro.steps || [])];
+      newSteps[stepIndex].modifiers = modifiers;
+      setNewMacro({ ...newMacro, steps: newSteps });
+    }
+  };
+
+  const handleDelayChange = (stepIndex: number, delay: number) => {
+    if (editingMacro) {
+      const newSteps = [...editingMacro.steps];
+      newSteps[stepIndex].delay = delay;
+      setEditingMacro({ ...editingMacro, steps: newSteps });
+    } else {
+      const newSteps = [...(newMacro.steps || [])];
+      newSteps[stepIndex].delay = delay;
+      setNewMacro({ ...newMacro, steps: newSteps });
+    }
+  };
+
+  const ErrorMessage = ({ error }: { error?: string }) => {
```

Comment:
Use the Fielderror in InputField.tsx

Reply:
converted everything over that i could see, lmk if i missed a shared component